tests/tool_excimer_sample_set_test.php

```diff
@@ -0,0 +1,127 @@
+ */
+class tool_excimer_sample_set_test extends excimer_testcase {
+
+    public function test_add_sample() {
+        $entries = [
+            $this->get_log_entry_stub(['a']),
+            $this->get_log_entry_stub(['b']),
+            $this->get_log_entry_stub(['c']),
+        ];
+
+        $set = new sample_set('a', 0);
+
+        foreach ($entries as $entry) {
+            $set->add_sample(($entry));
+        }
+
+        $this->assertEquals($entries, $set->samples);
+    }
+
+    public function test_filtering() {
+        $entries = [
+            $this->get_log_entry_stub(['a']),
+            $this->get_log_entry_stub(['b']),
+            $this->get_log_entry_stub(['c']),
+            $this->get_log_entry_stub(['d']),
+        ];
+
+        $set = new sample_set('a', 0);
+        $set->apply_doubling();
+
+        foreach ($entries as $entry) {
+            $set->add_sample(($entry));
+        }
+
+        $this->assertEquals([$entries[1], $entries[3]], $set->samples);
+
+        $set = new sample_set('a', 0);
+        $set->apply_doubling();
+        $set->apply_doubling();
+
+        foreach ($entries as $entry) {
+            $set->add_sample(($entry));
+        }
+
+        $this->assertEquals([$entries[3]], $set->samples);
+    }
+
+    public function test_stripping() {
+        $entries = [
+            $this->get_log_entry_stub(['a']),
+            $this->get_log_entry_stub(['b']),
+            $this->get_log_entry_stub(['c']),
+            $this->get_log_entry_stub(['d']),
+        ];
+
+        $set = new sample_set('a', 0);
+
+        foreach ($entries as $entry) {
+            $set->add_sample(($entry));
+        }
+
+        $set->apply_doubling();
+        $this->assertEquals([$entries[1], $entries[3]], $set->samples);
+
+        $set->apply_doubling();
+        $this->assertEquals([$entries[3]], $set->samples);
+    }
+
+    public function test_doubling() {
+        $entries1 = [
+            $this->get_log_entry_stub(['a']),
+            $this->get_log_entry_stub(['b']),
+            $this->get_log_entry_stub(['c']),
+            $this->get_log_entry_stub(['d']),
+            $this->get_log_entry_stub(['e']),
+            $this->get_log_entry_stub(['f']),
+        ];
+        $entries2 = [
+            $this->get_log_entry_stub(['g']),
+            $this->get_log_entry_stub(['h']),
+            $this->get_log_entry_stub(['i']),
+            $this->get_log_entry_stub(['j']),
+            $this->get_log_entry_stub(['k']),
+            $this->get_log_entry_stub(['l']),
+        ];
+
+        $set = new sample_set('a', 0, 4);
+
+        foreach ($entries1 as $entry) {
+            $set->add_sample(($entry));
+        }
+
+        $this->assertEquals([$entries1[1], $entries1[3], $entries1[5]], $set->samples);
+
+        foreach ($entries2 as $entry) {
+            $set->add_sample(($entry));
+        }
+
+        $this->assertEquals([$entries1[3], $entries2[1], $entries2[5]], $set->samples);
```

Comment:
Not being sure what's tested here aside from the fact it's testing some sort of doubling (function name).

I assume (after reading all the code) it's something to do with the limit set to 4, but at a glance I'm unsure how it's testing this or what the expectations should be. It would greatly help if there were comments here.